surface/include/pcl/surface/impl/convex_hull.hpp

```diff
@@ -182,106 +166,88 @@ pcl::ConvexHull<PointInT>::performReconstruction2D (PointCloud &hull, std::vecto
-  hull.points.resize (num_vertices);
-  memset (&hull.points[0], static_cast<int> (hull.points.size ()), sizeof (PointInT));
+    // 0 if no error from qhull or it doesn't find any vertices
+    if (qhull.qhullStatus() != 0 || qhull.vertexCount() == 0)
+    {
+      throw std::runtime_error(qhull.qhullMessage());
+    }
 
-  vertexT * vertex;
-  int i = 0;
+    // Qhull returns the area in volume for 2D
+    if (compute_area_)
+    {
+      total_area_ = qhull.volume();
+      total_volume_ = 0.0;
+    }
 
-  std::vector<std::pair<int, Eigen::Vector4f>, Eigen::aligned_allocator<std::pair<int, Eigen::Vector4f> > > idx_points (num_vertices);
-  idx_points.resize (hull.points.size ());
-  memset (&idx_points[0], static_cast<int> (hull.points.size ()), sizeof (std::pair<int, Eigen::Vector4f>));
+    std::vector<std::pair<int, Eigen::Vector4f>, Eigen::aligned_allocator<std::pair<int, Eigen::Vector4f> > > idx_points;
 
-  FORALLvertices
-  {
-    hull.points[i] = input_->points[(*indices_)[qh_pointid (vertex->point)]];
-    idx_points[i].first = qh_pointid (vertex->point);
-    ++i;
-  }
+    orgQhull::QhullVertexList vertexs = qhull.vertexList();
+    for(orgQhull::QhullVertexList::iterator it = vertexs.begin();it!=vertexs.end();it++)
+    {
+      hull.push_back(input_->points[(*indices_)[it->point().id()]]);
+      idx_points.push_back(std::make_pair(it->point().id(), Eigen::Vector4f()));
+    }
 
-  // Sort
-  Eigen::Vector4f centroid;
-  pcl::compute3DCentroid (hull, centroid);
-  if (xy_proj_safe)
-  {
-    for (size_t j = 0; j < hull.points.size (); j++)
+    // Sort
+    Eigen::Vector4f centroid;
+    pcl::compute3DCentroid (hull, centroid);
+    if (xy_proj_safe)
     {
-      idx_points[j].second[0] = hull.points[j].x - centroid[0];
-      idx_points[j].second[1] = hull.points[j].y - centroid[1];
+      for (size_t j = 0; j < hull.points.size (); j++)
+      {
+        idx_points[j].second[0] = hull.points[j].x - centroid[0];
+        idx_points[j].second[1] = hull.points[j].y - centroid[1];
+      }
     }
-  }
-  else if (yz_proj_safe)
-  {
-    for (size_t j = 0; j < hull.points.size (); j++)
+    else if (yz_proj_safe)
     {
-      idx_points[j].second[0] = hull.points[j].y - centroid[1];
-      idx_points[j].second[1] = hull.points[j].z - centroid[2];
+      for (size_t j = 0; j < hull.points.size (); j++)
+      {
+        idx_points[j].second[0] = hull.points[j].y - centroid[1];
+        idx_points[j].second[1] = hull.points[j].z - centroid[2];
+      }
     }
-  }
-  else if (xz_proj_safe)
-  {
-    for (size_t j = 0; j < hull.points.size (); j++)
+    else if (xz_proj_safe)
     {
-      idx_points[j].second[0] = hull.points[j].x - centroid[0];
-      idx_points[j].second[1] = hull.points[j].z - centroid[2];
+      for (size_t j = 0; j < hull.points.size (); j++)
+      {
+        idx_points[j].second[0] = hull.points[j].x - centroid[0];
+        idx_points[j].second[1] = hull.points[j].z - centroid[2];
+      }
```

Comment:
You can apply the similar strategy pointed before for this loop. You already have coefficients c0 and c1 available.